apis/_codegen/fetch-apis.go

```diff
@@ -1,99 +1,235 @@
-			os.Exit(1)
-		}
-		// Test that we can parse the JSON schema (otherwise it's invalid)
-		var i interface{}
-		if json.Unmarshal(res.Body, &i) != nil {
-			fmt.Println("Failed to parse: ", url, " error: ", err)
-			os.Exit(1)
-		}
-		schemas[url] = string(res.Body)
+
+		urls[url] = true
+		wg.Add(1)
+		go func() {
+			s := fetchSchema(g, url)
+
+			mutex.Lock()
+			schemas[url] = s
+			mutex.Unlock()
+			wg.Done()
+		}()
 	}
 	for _, s := range services {
 		for _, e := range s.Entries {
-			if e.Input != "" {
-				fetchSchema(e.Input)
-			}
-			if e.Output != "" {
-				fetchSchema(e.Output)
-			}
+			addSchema(e.Input)
+			addSchema(e.Output)
 		}
 	}
+	wg.Wait()
 
-	code := "package apis\n"
-	code += "\n"
-	code += "import \"github.com/taskcluster/taskcluster-cli/apis/definitions\"\n"
-	code += "\n"
-	code += "var services = " + fmt.Sprintf("%#v", services)
-	code += "\n"
-	code += "var schemas = " + fmt.Sprintf("%#v", schemas)
-	code += "\n"
+	gen.Print("var schemas = ")
+	gen.PrintComposite(schemas, true)
+	gen.Print("\n")
 
-	source, err := format.Source([]byte(code))
+	// Format the output.
+	source, err := gen.Format()
 	if err != nil {
-		fmt.Println("go fmt, code generation failed, error: ", err)
-		os.Exit(1)
+		log.Println(gen.String())
+		log.Fatalln("error: go fmt, code generation failed: ", err)
 	}
 
 	if err := ioutil.WriteFile("services.go", source, 0664); err != nil {
-		fmt.Println("Failed to save services.go, error: ", err)
-		os.Exit(1)
+		log.Fatalln("error: failed to save services.go: ", err)
 	}
 }
+
+func fetchService(g *got.Got, name string, url string) definitions.Service {
+	fmt.Println(" - Fetching", name)
+	// Fetch reference
+	res, err := g.Get(url).Send()
+	if err != nil {
+		log.Fatalln("error: failed to fetch API ", name, ": ", err)
+	}
+	// Parse reference
+	var s definitions.Service
+	if err := json.Unmarshal(res.Body, &s); err != nil {
+		log.Fatalln("error: failed parse API ", name, ": ", err)
+	}
+	return s
+}
+
+func fetchSchema(g *got.Got, url string) string {
+	fmt.Println(" - ", url)
+	res, err := g.Get(url).Send()
+	if err != nil {
+		log.Fatalln("error: failed to fetch ", url, ": ", err)
+	}
+	// Test that we can parse the JSON schema (otherwise it's invalid)
+	var i interface{}
+	if err := json.Unmarshal(res.Body, &i); err != nil {
+		log.Fatalln("error: failed to parse ", url, ": ", err)
+	}
+	return string(res.Body)
+}
+
+// generator holds a buffer of the output that will be generated.
+type generator struct {
+	buf bytes.Buffer
+}
+
+func (g *generator) Write(p []byte) (n int, err error) {
+	return g.buf.Write(p)
+}
+
+// Printf prints the given format+args to the buffer.
+func (g *generator) Printf(format string, args ...interface{}) {
+	fmt.Fprintf(&g.buf, format, args...)
+}
+
+// Print prints the given a to the buffer.
+func (g *generator) Print(a ...interface{}) {
+	fmt.Fprint(&g.buf, a...)
+}
+
+func (g *generator) PrintComposite(data interface{}, sorted bool) {
```

Comment:
sure thing. I was thinking that eventually this method might be extracted into a separate package, but we can cross that bridge when/if we get there.